src/engine/persistence/game_state_repository.py

```diff
@@ -0,0 +1,59 @@
+import pickle
+
+from yaml import Loader, dump, load
+
+from src.config import SAVE_FILE_DIRECTORY
+from src.engine.guild import Guild
+from src.engine.init_engine import eng
+from src.engine.persistence.loaders import GameStateLoaders
+
+
+class GameStateRepository:
+    @classmethod
+    def save_file_path(cls, slot: int, fmt: str = "pikl") -> tuple[str, ...]:
+        match fmt:
+            case "pikl":
+                return str(SAVE_FILE_DIRECTORY / f"save_{slot}.pikl")
+            case "yaml":
+                return str(SAVE_FILE_DIRECTORY / f"save_{slot}.yaml")
+            case _:
+                raise ValueError(
+                    f"Unsupported format. Got {fmt=}, expected: pikl / yaml"
+                )
+

```

Comment:
Just wee adjustment for brevity/extensibility.

```suggestion

from enum import Enum

class Format(Enum):
    YAML = "yaml"
    PICKLE = "pikl"
    
    def dumper(self):
        return {
            Format.YAML: dump,
            Format.PICKLE: pickle.dump,
        }[self]
        
    def loader(self):
        return {
            Format.YAML: lambda file: load(file, Loader),
            Format.PICKLE: pickle.load,
        }[self]
    
    def dump(self, data, path):
        with open(path, "wb+") as save_file:
            self.dumper(data, save_file)
            
     def load(self, path) -> dict:
         with open(path, "r") as save_file:
            state = self.loader(save_file)
            
         return state
        

class GameStateRepository:
    @classmethod
    def save_file_path(cls, slot: int, fmt: Format = Format.PICKLE) -> tuple[str, ...]:
        if not isinstance(fmt, Format):
            raise TypeError(f"Unrecognised format {fmt}")
        return str(SAVE_FILE_DIRECTORY / f"save_{slot}.{fmt.value}")
    
    @classmethod
    def load(cls, slot, fmt=Format.PICKLE):
        if not isinstance(fmt, Format):
            raise TypeError(f"Unrecognised format {fmt}")
            
        return GameStateLoaders.guild_from_dict(fmt.load(self.save_file_path(slot, fmt)))

```

Reply:
The idea being if you add another format you're forced to also create a load and dump function for those formats. Then there is only one save/load function

Reply:
just to be clear this is gold-plating, definitely non-essential